tests/blockchain/test_blockchain.py

```diff
@@ -2019,7 +2026,7 @@ async def test_cost_exceeds_max(self, empty_blockchain):
         blocks = bt.get_consecutive_blocks(
             1, block_list_input=blocks, guarantee_transaction_block=True, transaction_data=tx
         )
-        assert (await b.receive_block(blocks[-1]))[1] == Err.INVALID_BLOCK_COST
+        assert (await b.receive_block(blocks[-1]))[1] in [Err.BLOCK_COST_EXCEEDS_MAX, Err.INVALID_BLOCK_COST]
```

Comment:
the exact error code changes a bit here. The distinction between exceeding cost, and the cost mismatching what the header declared is a bit fuzzy. As the cost declared in the header is *used* as the limit, and if the program exceeds it, it both exceeds the cost *and* is invalid. Either way, it fails, which is what matters.

Reply:
The cost at the header might be larger than our max block cost constant, does it check for that ?

Reply:
I would expect the header validation would reject any block whose header declares the cost is higher than the max allowed cost. This patch doesn't really affect that part though. We limit the allowed cost by the declared cost (or the max block cost, whichever is smaller) and run the program. Then, while parsing the conditions, we further limit the cost by what's left. The cost-left is maintained as conditions are parsed, allowing us to fail immediately, during parsing, if the conditions cause the cost to exceed max.

That case will just fail with `BLOCK_COST_EXCEEDS_MAX`, because the max is set based on the block header.

Previously we would run the program, parse the conditions and *then* check to see if their cost is the same or different from the declared cost in the block header. The new behaviour lets us fail earlier.